src/java/org/apache/cassandra/index/sai/disk/v1/postings/PostingsReader.java

```diff
@@ -0,0 +1,395 @@
+
+import org.apache.cassandra.index.sai.disk.PostingList;
+import org.apache.cassandra.index.sai.disk.v1.DirectReaders;
+import org.apache.cassandra.index.sai.disk.v1.LongArray;
+import org.apache.cassandra.index.sai.metrics.QueryEventListener;
+import org.apache.cassandra.index.sai.utils.SeekingRandomAccessInput;
+import org.apache.lucene.index.CorruptIndexException;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.store.RandomAccessInput;
+
+
+/**
+ * Reads, decompresses and decodes postings lists written by {@link PostingsWriter}.
+ *
+ * Holds exactly one posting block in memory at a time. Does binary search over skip table to find a postings block to
+ * load.
+ */
+@NotThreadSafe
+public class PostingsReader implements OrdinalPostingList
+{
+    protected final IndexInput input;
+    private final int blockSize;
+    private final long numPostings;
+    private final LongArray blockOffsets;
+    private final LongArray blockMaxValues;
+    private final SeekingRandomAccessInput seekingInput;
+    private final QueryEventListener.PostingListEventListener listener;
+
+    // TODO: Expose more things through the summary, now that it's an actual field?
+    private final BlocksSummary summary;
+
+    private int postingsBlockIdx;
+    private int blockIdx; // position in block
+    private long totalPostingsRead;
+    private long actualSegmentRowId;
+
+    private long currentPosition;
+    private DirectReaders.Reader currentFORValues;
+    private long postingsDecoded = 0;
+
+    @VisibleForTesting
+    public PostingsReader(IndexInput input, long summaryOffset, QueryEventListener.PostingListEventListener listener) throws IOException
+    {
+        this(input, new BlocksSummary(input, summaryOffset, () -> {}), listener);
+    }
+
+    public PostingsReader(IndexInput input, BlocksSummary summary, QueryEventListener.PostingListEventListener listener) throws IOException
+    {
+        this.input = input;
+        this.seekingInput = new SeekingRandomAccessInput(input);
+        this.blockOffsets = summary.offsets;
+        this.blockSize = summary.blockSize;
+        this.numPostings = summary.numPostings;
+        this.blockMaxValues = summary.maxValues;
+        this.listener = listener;
+
+        this.summary = summary;
+
+        reBuffer();
+    }
+
+    @Override
+    public long getOrdinal()
+    {
+        return totalPostingsRead;
+    }
+
+    interface InputCloser
+    {
+        void close() throws IOException;
+    }
+
+    public static class BlocksSummary
+    {
+        final int blockSize;
+        final int numPostings;
+        final LongArray offsets;
+        final LongArray maxValues;
+
+        private final InputCloser runOnClose;
+
+        public BlocksSummary(IndexInput input, long offset) throws IOException
+        {
+            this(input, offset, input::close);
+        }
+
+        public BlocksSummary(IndexInput input, long offset, InputCloser runOnClose) throws IOException
```

Comment:
We can just use `Closeable`, which is implemented by `IndexInput`. That would save us the `InputCloser` interface. In fact, the `runOnClose` argument can be just a boolean, unless we have plans for something different than just closing the `IndexInput` passed as argument.